patriciaTree.ml

```diff
@@ -588,6 +594,163 @@ module WeakSetNode(Key:sig type 'a t end)(* :NODE *) = struct
 
 end
 
+let sdbm x y = y + (x lsl 16) + (x lsl 6) - x
+(** Combine two numbers into a new hash *)
+
+module HashconsedNode(Key:HETEROGENEOUS_KEY)(Value:VALUE): HASH_CONSED_NODE
+  with type 'key key = 'key Key.t
+   and type ('key,'map) value = ('key,'map) Value.t
+= struct
+
+  type 'a key = 'a Key.t
+  type ('key,'map) value = ('key,'map) Value.t
+
+  type 'map view =
+    | Empty: 'map view
+    | Branch: { prefix:intkey; branching_bit:mask; tree0:'map t; tree1:'map t } -> 'map view
+    | Leaf: { key:'key key; value:('key,'map) value } -> 'map view
+  and 'map t =
+    | NEmpty: 'map t
+    | NBranch: { prefix:intkey; branching_bit:mask; tree0:'map t; tree1:'map t; id:int } -> 'map t
+    | NLeaf: { key:'key key; value:('key,'map) value; id:int } -> 'map t
+
+  let view = function
+    | NEmpty -> Empty
+    | NBranch{prefix;branching_bit;tree0;tree1;_} -> Branch{prefix;branching_bit;tree0;tree1}
+    | NLeaf{key;value;_} -> Leaf{key;value}
+
+  let get_id = function
+    | NEmpty -> 0
+    | NBranch{ id; _ } -> id
+    | NLeaf{ id; _ } -> id
+
+  let count = ref 1 (** Start at 1 as we increment in post *)
+
+  module HashArg = struct
+    type 'a map = 'a t
+    type t = Exi : 'a map -> t [@@unboxed]
+    let equal (Exi a) (Exi b) = match a, b with
+      | NEmpty, NEmpty -> true
+      | NLeaf{key=key1;value=value1;_}, NLeaf{key=key2;value=value2;_} ->
+        begin match Key.polyeq key1 key2 with
+        | Eq -> value1 == Obj.magic value2
+        | Diff -> false
+        end
+      | NBranch{prefix=prefixa;branching_bit=branching_bita;tree0=tree0a;tree1=tree1a;_},
+        NBranch{prefix=prefixb;branching_bit=branching_bitb;tree0=tree0b;tree1=tree1b;_} ->
+        prefixa == prefixb && branching_bita == branching_bitb &&
+        get_id tree0a = get_id tree0b && get_id tree1a = get_id tree1b
```

Comment:
I think we can use == instead of get_id here

Reply:
Unfortunately not, because these values don't have the same type... This is also why I need an `Obj.magic` in the leaf case.

However, the more I think about it, the more I believe we should get rid of arbitrary value maps for hash-consing. It makes little sense to hashcons maps from `'a key` to `('a, 'b) value` together with map from `'a key` to `('a, 'c) value`. So I'll probably change it to use a `'a value` type instead of the more generic `('a, 'b) value` type. 

Otherwise we might have problems with different types having equal values.